src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs

```diff
@@ -138,25 +138,27 @@ internal static partial class Extensions
 
                     return result;
                 }
-            }
-            catch (Exception ex) when (IsCancelled(ex))
-            {
-                cancellationToken.ThrowIfCancellationRequested();
-                throw;
+                catch (Exception ex) when (IsCancelled(ex, mergedCancellation.Token))
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    throw;
+                }
             }
         }
 
-        private static bool IsCancelled(Exception ex)
+        private static bool IsCancelled(Exception ex, CancellationToken token)
         {
-            // object disposed exception can be thrown from StreamJsonRpc if JsonRpc is disposed in the middle of read/write.
-            // the way we added cancellation support to the JsonRpc which doesn't support cancellation natively
-            // can cause this exception to happen. newer version supports cancellation token natively, but
-            // we can't use it now, so we will catch object disposed exception and check cancellation token
-            if (ex is ObjectDisposedException || ex is OperationCanceledException)
+            // check whether we are in cancellation mode
+            if (token.IsCancellationRequested)
```

Comment:
yep. that is the purpose of this change. when cancellation is raised, various exception can happen based on what was happening when we forcefully closed named pipe stream.